README.md
# rfc822-eld

PHP email address, local part and domain validator based on RFCs 822, 2822, 5322

This modified version of the code allows for the partial checking of email addresses, local parts and domains.
By default the code works as before but if the extra options are used, the code can be made to check whether some
string is just a local part or just a domain. That means that if one provides a full email address while asking to
perform a domain check, the could will return an error because a full email addresss is not a domain. Likewise the
test of a local part on a full email address will return an error also. Also local parts may well not be domain names
and vice versa. The PHP built-in FILTER_VALIDATE_DOMAIN was tested but did not appear to produce the right answers at
all times.


CHANGELOG
rfc822-eld version 5012156+1 (14.12.2022)
-------------------------------

  - Enhanced the main function so that it can check full email addresses, or just local parts and domain names.
    If it has been asked to check whether  a string is a full address, having been provided with just a local
    part or domain, it will return an error. If the email address itself is syntactically incorrect, it will
    also produce an error. Similarly, it can check whether a string might be a suitable local part or a suitable
    domain name. By default the function will behave as before. 

  - Fixed the code that drives the comment checking so that it fails if a comment is used and the flag is disabled.

  - Fixed the testing and home page php code so that it works in today's world.


tests/cli.php
<?php
	include('runner.php');

	$ran		= 0;
	$public_passed	= 0;
	$strict_passed	= 0;

	foreach ($tests as $test){

		$ran++;

		if ($test['expected'] == $test['result_public']){
			$public_passed++;
		} else{
			print_r($test);
		}

		if ($test['expected'] == $test['result_strict']){
			$strict_passed++;
		} else{
			print_r($test);
		}

	}

	echo "Public passed $public_passed of $ran\n";
	echo "Strict passed $strict_passed of $ran\n";
?>


tests/index.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html><head>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8" />
<title>Cal Henderson - RFC-compliant email address validation</title>

<style type="text/css">

body {
	font-family:Calibri,Tahoma,Verdana,Arial,Helvetica,sans serif;
	color: #264E62;
}

p,h1,h2,h3,h4,h5,h6 {
	font-family:inherit;
	color:inherit;
	padding-left:7px;
	padding-right:7px;
	margin-top:0px;
	margin-left:8px;
	margin-right:8px;
}

a:link {
	text-decoration: none;
	color: #266139;
}
a:visited {
	text-decoration: none;
	color: #34854E;
}
a:hover {
	text-decoration: underline;
	color: #7EA58A;
}
a:active {
	text-decoration: none;
	color: #264E62;
}
a img {
	border-style: none;
	vertical-align:baseline;
}

h1, h2, h3, h4, h5, h6 {
	margin-bottom: 1em;
}

h1 {
	color: #346A85;
}

.isemail {
	font-size:12px;
	clear:left;
	margin:0;
	padding:0;
}

.isemail_address {
	line-height:12pt;
	float:left;
	overflow:hidden;
	width:250px;
	margin:0 0 0 8px;
	padding:3px;
}

.isemail_result {
	float:left;
	width:80px;
	margin:0 3px 0 3px;
	padding:3px;
	text-align: center;
}

.isemail_expected			{background-color:white;}
.isemail_unexpected			{background-color:#FFCCCC;}
.isemail_header				{background-color:#CCCCCC;margin-bottom:3px;}

.isemail_tooltip			{position:relative;}
.isemail_tooltip:hover span	{display: block;}
.isemail_tooltip span {
	display: none;
	position: absolute;
	top: 20px;
	left: 10px;
	padding: 5px;
	z-index: 100;
	background-color: #323232;
	color: #fff;
	width:auto;
	-moz-border-radius: 5px; /* this works only in camino/firefox */
	-webkit-border-radius: 5px; /* this is just for Safari */
}

</style>

</head>

<body>

<h1>RFC-compliant email address validation</h1>

<?php
	include('runner.php');

#echo "<pre>";
#echo HtmlSpecialChars(print_r($tests, 1));
#echo HtmlSpecialChars(print_r($totals, 1));
#echo "</pre>";
?>

<p>
	This test suite comes from <a href="http://www.dominicsayers.com/isemail/">Dominic Sayers</a> and is a mix of RFC examples and examples from other validators.
</p>
<p>
	The test suite is currently allowing addresses that are unroutable on the public Internet, such as <code>first.last@example.123</code>.
	This behavior is disabled by default, but can be switched on; the tests are run in both modes for comparison.
</p>

<div class="isemail isemail_tooltip">
	<p class="isemail_address"><br />&nbsp;</p>

	<p class="isemail_result isemail_header">Expected</p>
	<p class="isemail_result isemail_header">Strict Mode</p>
	<p class="isemail_result isemail_header">Public Mode</p>
</div>

<div class="isemail isemail_tooltip">
	<p class="isemail_address">Percent correct</p>
	<p class="isemail_result isemail_expected">-</p>
	<p class="isemail_result isemail_expected"><?php echo floor(100 * $totals['strict'] / $totals['all'])?>%</p>
	<p class="isemail_result isemail_expected"><?php echo floor(100 * $totals['public'] / $totals['all'])?>%</p>
</div>

<div class="isemail isemail_tooltip">
	<p class="isemail_address">Number correct</p>
	<p class="isemail_result isemail_expected">-</p>
	<p class="isemail_result isemail_expected"><?php echo $totals['strict']?> / <?php echo $totals['all']?></p>
	<p class="isemail_result isemail_expected"><?php echo $totals['public']?> / <?php echo $totals['all']?></p>
</div>


<?php foreach ($tests as $test){ ?>

<div class="isemail isemail_tooltip">
	<span>
		Test # <?php echo $test['id']?><br />
		<strong><?php echo show_escapes(HtmlSpecialChars($test['address']))?></strong><br />
		Expected result: <?php echo is_valid($test['expected'])?><br />
<?php if ($test['comment']){ ?>
		Comment: <?php echo HtmlSpecialChars($test['comment'])?><br />
<?php } ?>
		Source: <?php echo HtmlSpecialChars($test['source'])?>
	</span>
	<p class="isemail_address"><nobr><a href="<?php echo HtmlSpecialChars($test['sourcelink'])?>" target="_blank"><?php echo show_escapes(HtmlSpecialChars($test['address']))?></a></nobr></p>
	<p class="isemail_result isemail_expected"><?php echo is_valid($test['expected'])?></p>

	<p class="isemail_result isemail_<?php echo $test['result_strict']==$test['expected']?'':'un'?>expected"><?php echo is_valid($test['result_strict'])?></p>
		<p class="isemail_result isemail_<?php echo $test['result_public']==$test['expected']?'':'un'?>expected"><?php echo is_valid($test['result_public'])?></p>
</div>

<?php } ?>

<p style="clear: both; padding-bottom: 50px;">&nbsp;</p>

</body>
</html>


tests/reader.php
<?php 
	error_reporting(30719 | 2048); # E_ALL | E_STRICT
	ini_set("display_errors", 1);


	#
	# read the tests (PHP 4 compatible!)
	#

	$test = array();
	$tests = array();
	$glossary = array();
	$last_key = '?';
	$last_glossary_id = '?';
	
	$parser = xml_parser_create();

	function start($parser, $element_name, $element_attrs){
		$GLOBALS['last_key'] = StrToLower($element_name);

		if ($element_name == 'GLOSSARY'){
			$GLOBALS['last_glossary_id'] = $element_attrs['ID'];
		}
	}

	function stop($parser, $element_name){
		$GLOBALS['last_key'] = '?';
		if ($element_name == 'TEST'){
			$GLOBALS['tests'][$GLOBALS['test']['id']] = $GLOBALS['test'];
			$GLOBALS['test'] = array();
		}
		if ($element_name == 'TAG'){
			$GLOBALS['test']['tags'][] = $GLOBALS['test']['tag'];
			unset($GLOBALS['test']['tag']);
		}
		if ($element_name == 'GLOSSARY'){
			$GLOBALS['glossary'][$GLOBALS['last_glossary_id']] = $GLOBALS['test'];
			$GLOBALS['test'] = array();
		}
	}

	function char($parser, $chr){
		$chr = str_replace('\\0', "\0", $chr);
		if ($GLOBALS['last_key'] == 'tests') return;
		if ($GLOBALS['last_key'] == 'test') return;
		if ($GLOBALS['last_key'] == 'glossary') return;
		if ($GLOBALS['last_key'] == '?') return;

		if (isset($GLOBALS['test'][$GLOBALS['last_key']])){

			$GLOBALS['test'][$GLOBALS['last_key']] .= $chr;
		}else{
			$GLOBALS['test'][$GLOBALS['last_key']] = $chr;
		}
	}

	xml_set_element_handler($parser, "start", "stop");
	xml_set_character_data_handler($parser, "char");
	xml_parser_set_option($parser, XML_OPTION_SKIP_WHITE, 1);

	$fp = fopen("tests.xml", "r");

	while ($data = fread($fp, 4096)){
		xml_parse($parser, $data, feof($fp)) or die (sprintf("XML Error: %s at line %d", xml_error_string(xml_get_error_code($parser)), xml_get_current_line_number($parser)));
	}
	xml_parser_free($parser);


	#
	# perform some type conversion
	#

	foreach ($tests as $k => $v){

		$tests[$k]['valid'] = $tests[$k]['valid'] == 'true' ? true : false;
		$tests[$k]['warning'] = $tests[$k]['warning'] == 'true' ? true : false;
		$tests[$k]['id'] = intval($tests[$k]['id']);
	}

?>


demo.php
<?php
	$title = "RFC 822 Demo";
	$no_homelink = 1;
	include('../../head.txt');

	include('rfc822.php');
?>

<p class="homelink"><a href="/rfc822-eld/">&laquo; Back to project homepage</a></p>

<h1>RFC 822 Demo</h1>

<p>Enter an email address to test:</p>

<form action="demo.php" method="get">
<input type="text" name="e" value="<?php echo HtmlSpecialChars($_GET['e'])?>" style="width: 400px" />
<input type="submit" value="Go" />
</form>

<?php if (strlen($_GET['e'])){ ?>

<hr />

<p><b>Address:</b> <?php echo HtmlSpecialChars($_GET['e'])?></p>
<p><b>Result (Public Internet):</b>
<?php if (is_valid_email_address($_GET['e'],array('public_internet' => true))){ ?>
	<span style="color: green">Valid</span>
<?php }else{ ?>
	<span style="color: red">Invalid</span>
<?php }?>
</p>
<p><b>Result (Strict):</b>
<?php if (is_valid_email_address($_GET['e'],array('public_internet' => false))){ ?>
	<span style="color: green">Valid</span>
<?php }else{ ?>
	<span style="color: red">Invalid</span>
<?php }?>
</p>

<?php } ?>


<?php
	include('../../foot.txt');
?>


index.php
<?php
	$title = "RFC 822 Email Address Parser in PHP";
	include('../../head.txt');
?>

<h1>RFC 822  Email Address Parser in PHP</h1>


<h2>Source code</h2>

<ul>
	<li> <a href="https://github.com/2024sight/rfc822-eld/blob/master/rfc822.php">rfc822.php</a> - Email Address Parser </li>
</ul>

<h2>Demo</h2>

<p>There is an <a href="/rfc822-eld/demo.php">online interactive demo</a> if you'd just like to give it a try.</p>

<h2>Tests</h2>

<p>The <a href="tests/">test suite</a> shows parser's results, based on these <a href="tests/tests.xml">test definitions</a>. These are borrowed from
<a href="http://www.dominicsayers.com/isemail/">Dominic Sayers</a> who has a similar parser.</p>


<h2>The RFCs</h2>

<p>This library was named back when there was only one RFC for email addresses; there are now lots, so it would be better named RFC 822/2822/5322 at the least. These are the most relevant 
ones:</p>

<dl>
	<dt><a href="http://www.faqs.org/rfcs/rfc821.html">RFC 821 - Simple Mail Transfer Protocol</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=821">Errata</a>)</dt>
	<dd>The original SMTP RFC</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc822.html">RFC 822 - Standard for the Format of ARPA Internet Text Messages</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=822">Errata</a>)</dt>
	<dd>The original 'email' RFC</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc1035.html">RFC 1035 - Domain names - implementation and specification</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=1035">Errata</a>)</dt>
	<dd>The old domains RFC</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc1123.html">RFC 1123 - Requirements for Internet Hosts - Application and Support</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=1123">Errata</a>)</dt>
	<dd>An update to RFC 1035</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc2821.html">RFC 2821 - Simple Mail Transfer Protocol</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=2821">Errata</a>)</dt>
	<dd>SMTP contains some address limits not in RFC 2822</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc2822.html">RFC 2822 - Internet Message Format</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=2822">Errata</a>)</dt>
	<dd>Superceeds RFC 822</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc3696.html">RFC 3696 - Application Techniques for Checking and Transformation of Names</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=3696">Errata</a>)</dt>
	<dd>An informative RFC that clarifies some rules (and muddies others)</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc4291.html">RFC 4291 - IP Version 6 Addressing Architecture</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=4291">Errata</a>)</dt>
	<dd>Some useful details about the horrors of IPv6</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc5321.html">RFC 5321 - Simple Mail Transfer Protocol</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=5321">Errata</a>)</dt>
	<dd>Superceeds RFC 2821 (this is the latest SMTP RFC)</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc5322.html">RFC 5322 - Internet Message Format</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=5322">Errata</a>)</dt>
	<dd>Superceeds RFC 2822 (this is the latest email RFC)</dd>

	<dt><a href="http://www.faqs.org/rfcs/rfc5952.html">RFC 5952 - A Recommendation for IPv6 Address Text Representation</a> (<a href="http://www.rfc-editor.org/errata_search.php?rfc=5952">Errata</a>)</dt>
	<dd>Superceeds RFC 4291 (this is the latest IPv6 RFC)</dd>
</dl>

<p>Reading the errata is pretty important, since some of the examples and even the EBNF are wrong in the original RFCs.</p>


<h2>Copyright</h2>

<p>By Cal Henderson &lt;cal@iamcal.com&gt;</p>
<p>
	This code is dual licensed:<br />
	Creative Commons Attribution-ShareAlike 2.5 License - <a href="http://creativecommons.org/licenses/by-sa/2.5/">http://creativecommons.org/licenses/by-sa/2.5/</a><br />
	GNU General Public License v3 - <a href="http://www.gnu.org/copyleft/gpl.html">http://www.gnu.org/copyleft/gpl.html</a><br />
</p>
<p>
	If you require the code to be released under a different license, please contact the author.
</p>


<h2>Limitations</h2>

<p>
	The code only verifies that the email address matches the various RFC specs.
	<b>This does not mean it's a valid Internet email address!</b>
	For an email address to be valid on the Internet, the domain part must be a valid domain name, be resolvable and have an MX.
	The code will identify the address "<code>foo@bar.baz</code>" as valid, even though we konw that there's no such domain as <code>bar.baz</code>.
	If you want to check that it's valid, fetching the MX for the domain is a good start.
	Connecting to the MX to verify it's a mail server is even better.
</p>


<h2>Extras</h2>

<p>Tim Fletcher has translated the function to ruby and python: <a href="http://tfletcher.com/lib/">http://tfletcher.com/lib/</a>. </p>

<p>Anton Hofland added the partial test functionality to just check local part and domain instead of full email addresses. </p>

<p>
	A fullly unpacked version of the underlying regular expression can be seen <a href="full_regexp.txt">here</a>. It's huge.
</p>

<p>
	It's been said that it's impossible to parse email addresses using regular expressions alone. This is somewhat true.
	If you allow comments in email addresses, then nested comments cannot be matched with a single regexp - a simple loop applying a reducing regexp first is needed.
	Aside from that, this library uses some post-match checks instead of rolling everything into one regexp.
	This is not because it wouldn't be possible, but because it would make it <i>huge</i> - the number of IPv6 permutations alone would probably double the size.
	Aside from the practicality, it seems entirely possible to boil it down to a single regexp.
	However, the <a href="http://www.whatwg.org/specs/web-apps/current-work/multipage/states-of-the-type-attribute.html#e-mail-state">one used for HTML5</a> is not even close :(
</p>


<?php
	include('../../foot.txt');
?>
